File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "BOJ/input.txt";
let [n, ...input] = fs.readFileSync(filePath).toString().trim().split("\n");

input = input.map((v) => v.split(" ").map(Number));
const dp = [input[0]];

for (let i = 1; i < input.length; i++) {
  let arr = [];

  for (let j = 0; j < input[i].length; j++) {
    let value = 0;
    if (j === 0) {
      value = input[i][0] + dp[i - 1][0];
    } else if (j === input[i].length - 1) {
      value = input[i][j] + dp[i - 1][dp[i - 1].length - 1];
    } else {
      value = Math.max(
        input[i][j] + dp[i - 1][j - 1],
        input[i][j] + dp[i - 1][j]
      );
    }
    arr.push(value);
  }

  dp[i] = arr;
}

console.log(Math.max(...dp[dp.length - 1]));

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
import math

def solution(triangle):
    dp = [[] for _ in range(len(triangle))]
    for i in range(len(triangle)):
        dp[i] = [0] * len(triangle[i])
        
    dp[0][0] = triangle[0][0]
    dp[1][0] = dp[0][0] + triangle[1][0]
    dp[1][1] = dp[0][0] + triangle[1][1]
#     dp[2][0] = dp[1][0] + triangle[2][0]
#     dp[2][1] = max(dp[1][0] + triangle[2][1], dp[1][1] + triangle[2][1])
#     dp[2][2] = dp[1][1] + triangle[2][2]
    
#     dp[3][0] = dp[2][0] + triangle[3][0]
#     dp[3][3] = dp[2][2] + triangle[3][3]
    for j in range(2, len(triangle)):
        for k in range(0, len(triangle[j])):
            if k == 0:
                dp[j][k] = dp[j - 1][k] + triangle[j][k]
            elif k == len(triangle[j]) - 1:
                dp[j][k] = dp[j - 1][k - 1] + triangle[j][k]
            else:
                dp[j][k] = max(dp[j - 1][k - 1] + triangle[j][k], dp[j - 1][k] + triangle[j][k])
    
    return max(dp[len(triangle) - 1])